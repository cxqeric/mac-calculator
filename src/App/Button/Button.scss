input {
  border: 1px solid #808080;
  outline: none;
  @include font(7em, 200);

  &.Digit, &.Dot {
    background: #E0E0E0;
    @include font(6em, 200);
  }
  &.span-2 {
    padding-right: 23px;
  }

  &.Binary, &.Equal {
    background: #F5923E;
    color: white;
  }

  &.Unary, &.Reset {
    background: #D6D6D6;
    @include font(5em, 200);
  }

}