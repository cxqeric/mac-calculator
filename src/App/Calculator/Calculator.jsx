import React from 'react';
import PropTypes from 'prop-types';

import {
  START,
  START_DECIMAL,
  CUMUL,
  CUMUL_DECIMAL,
  LOW,
  availableShortcuts,
  cumulate,
  multiply,
  getDataByValue,
  getDataByShortcut } from '../../utils';

const initialState = {
  inputs: [0],
  pendingOps: [],
  target: {
    index: 0,
    status: START
  },
  output: '0',
  initial: true
};

class Calculator extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ])
  };

  static defaultProps = {
    children: null
  };

  static CALCULATOR_CONTEXT = '__calculator__';

  static childContextTypes = {
    [Calculator.CALCULATOR_CONTEXT]: PropTypes.shape({
      initial: PropTypes.bool,
      output: PropTypes.string,
      handleInput: PropTypes.func
    })
  };

  state = { ...initialState };


  getChildContext = () => ({
    [Calculator.CALCULATOR_CONTEXT]: {
      initial: this.state.initial,
      output: this.state.output,
      handleInput: this.handleInput
    }
  });

  componentDidMount() {
    document.onkeydown = this.handleKeyPress;
  }


  handleKeyPress = ({ key: shortcut }) => {
    if (!availableShortcuts.includes(shortcut)) {
      return;
    }
    const { value } = getDataByShortcut({ shortcut });
    this.handleInput({ value });
  }

  handleInput = ({ value: targetValue }) => {
    const targetData = getDataByValue({ value: targetValue });
    return targetData ?
      this[`handle${targetData.type}`](targetData) :
      null;
  };

  handleDigit = ({ value: digit }) => {
    const target = { ...this.state.target };
    const inputs = [...this.state.inputs];
    const { index, status } = { ...target };
    const output = cumulate(inputs[index] || 0, digit, status);
    if (status === START) {
      target.status = CUMUL;
      if (index === 2 && typeof inputs[2] !== 'undefined') {
        Object.assign(inputs, { 0: inputs[1], 1: inputs[0] });
      }
    } else if (status === START_DECIMAL) {
      target.status = CUMUL_DECIMAL;
    }
    inputs[index] = output;
    this.setState({
      inputs, target, output, initial: false
    });
  };

  handleBinary = (operation) => {
    const target = { ...this.state.target };
    const pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    if (index === 0) {
      target.status = START;
      target.index = 1;
      pendingOps[0] = operation;
      return this.setState({
        pendingOps,
        target
      });
    }

    return operation.priority === LOW ?
      this.inputLowBinaryOp(operation) :
      this.inputHighBinaryOp(operation);
  };

  inputLowBinaryOp = (operation) => {
    const target = { ...this.state.target };
    let inputs = [...this.state.inputs];
    let pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    let { output } = { ...this.state };
    target.status = START;
    if (index === 1) {
      if (typeof inputs[1] !== 'undefined') {
        output = pendingOps[0].work(inputs[0], inputs[1]);
      }
      pendingOps[0] = operation;
    } else if (typeof inputs[2] !== 'undefined') {
      output = pendingOps[0].work(inputs[0], pendingOps[1].work(inputs[1], inputs[2]));
    } else {
      output = pendingOps[0].work(inputs[0], inputs[1]);
      inputs[1] = output;
      pendingOps = [operation];
    }
    inputs = [output];
    target.index = 1;
    this.setState({
      inputs,
      target,
      pendingOps,
      output,
      initial: false
    });
  };

  inputHighBinaryOp = (operation) => {
    const target = { ...this.state.target };
    let inputs = [...this.state.inputs];
    const pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    let { output } = { ...this.state };
    target.status = START;
    if (typeof inputs[index] === 'undefined') {
      pendingOps[index - 1] = operation;
    } else if (index === 1) {
      if (pendingOps[0].priority === LOW) {
        pendingOps[1] = operation;
        target.index = 2;
      } else {
        output = pendingOps[0].work(inputs[0], inputs[1]);
        inputs = [output];
        pendingOps[0] = operation;
        target.index = 0;
      }
    } else {
      output = pendingOps[1].work(inputs[1], inputs[2]);
    }
    this.setState({
      inputs,
      target,
      pendingOps,
      output,
      initial: false
    });
  };

  handleUnary = ({ key, work }) => {
    const target = { ...this.state.target };
    const inputs = [...this.state.inputs];
    const { index, status } = { ...target };
    let { output } = { ...this.state };
    if (index === 0 || status !== START) {
      output = work(inputs[index]);
      target.status = START;
    } else {
      output = key === '%' ?
        multiply(inputs[index - 1], work(inputs[index - 1])) :
        work(inputs[index - 1]);
      if (index === 1) {
        target.index = 0;
      }
    }
    inputs[index] = output;
    this.setState({
      inputs, target, output, initial: false
    });
  };

  handleDot = () => {
    const target = { ...this.state.target };
    const inputs = [...this.state.inputs];
    const { index, status } = { ...target };
    if (status === CUMUL_DECIMAL) {
      return;
    }
    target.status = START_DECIMAL;
    if (typeof inputs[index] === 'undefined') {
      inputs[index] = 0;
    }
    const output = `${inputs[index]}.`;
    this.setState({
      inputs, target, output, initial: false
    });
  };

  handleEqual = () => {
    const target = { ...this.state.target };
    let inputs = [...this.state.inputs];
    let pendingOps = [...this.state.pendingOps];
    const { index } = { ...target };
    let { output } = { ...this.state };

    target.status = START;
    switch (index) {
      case 0:
        if (typeof pendingOps[0] !== 'undefined') {
          output = pendingOps[0].work(inputs[0], inputs[1]);
        }
        break;
      case 1:
        if (typeof inputs[1] === 'undefined') {
          inputs = [inputs[0], inputs[0]];
        }
        output = pendingOps[0].work(inputs[0], inputs[1]);
        break;
      default:
        if (typeof inputs[2] !== 'undefined') {
          output = pendingOps[0].work(inputs[0], pendingOps[1].work(inputs[1], inputs[2]));
          inputs = [inputs[0], inputs[2]];
        } else {
          output = pendingOps[0].work(pendingOps[1].work(inputs[0], inputs[1]), inputs[1]);
        }
        pendingOps = [pendingOps[1]];
    }
    inputs[0] = output;
    target.index = 0;
    this.setState({
      inputs,
      target,
      output,
      pendingOps
    });
  };

  handleReset = () => this.setState(initialState);

  render() {
    return (
      <div className="calculator">
        { this.props.children }
      </div>
    );
  }
}


export default Calculator;
