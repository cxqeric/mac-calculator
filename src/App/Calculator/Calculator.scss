.calculator {
  display: flex;
  flex-flow: column nowrap;
  justify-content: flex-start;
  width: 100%;
  height: 100%;
  color: white;

  .buttons {
    flex-basis: 80%;
    display: flex;
    flex-wrap: wrap;
    width: 100%;
  }  
}