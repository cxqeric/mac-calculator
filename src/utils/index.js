import { BigNumber } from 'bignumber.js';

BigNumber.config({
  EXPONENTIAL_AT: 17,
  DECIMAL_PLACES: 15
});

/**
 * constants
 */
export const START = 'START';
export const START_DECIMAL = 'START_DECIMAL';
export const CUMUL = 'CUMUL';
export const CUMUL_DECIMAL = 'CUMUL_DECIMAL';
export const LOW = 'LOW';
export const HIGH = 'HIGH';
export const NOT_A_NUMBER = 'Not a number';

/**
 *
 * @param value1
 * @param value2
 * @param type
 * @returns  concatinated value
 */
export const concat = (value1, value2, type) => {
  if (type === START) {
    return value2;
  }

  try {
    const bigNumber1 = new BigNumber(value1);
    const bigNumber2 = new BigNumber(value2);
    if (!bigNumber1.isFinite() || !bigNumber2.isFinite()) {
      throw new Error('value1 and value2 must both be finite values.');
    }

    const formattedValue1 = bigNumber1.toPrecision();
    const formattedValue2 = bigNumber2.toPrecision();

    if (type === START_DECIMAL) {
      return `${formattedValue1}.${formattedValue2}`;
    }
    if (type === CUMUL_DECIMAL && value1.split('.')[1] === '0') {
      return `${formattedValue1}${formattedValue2}`;
    }
    return `${formattedValue1}${formattedValue2}`;
  } catch (e) {
    return NOT_A_NUMBER;
  }
};

/**
 * @param a
 * @param operation
 * @param b
 * @returns  result of performing safely operation on a and b safely
 */
const safeOperation = (a, operation, b) => {
  try {
    const A = new BigNumber(a);
    const B = new BigNumber(b);
    const result = A[operation](B).div(1);
    if (!A.isFinite() || !B.isFinite() || !result.isFinite()) {
      throw new Error();
    } else {
      return result.toString();
    }
  } catch (e) {
    return NOT_A_NUMBER;
  }
};

/**
 * Binary Operations
 */
export const add = (a, b) => safeOperation(a, 'add', b);
export const subtract = (a, b) => safeOperation(a, 'sub', b);
export const multiply = (a, b) => safeOperation(a, 'mul', b);
export const divide = (a, b) => safeOperation(a, 'div', b);

/**
 * Unary Operations
 */
export const negate = a => multiply(a, -1);
export const percent = a => multiply(a, 0.01);

/**
 * digits inputs
 */
export const digits = [7, 8, 9, 4, 5, 6, 1, 2, 3, 0]
  .map(digit => ({
    value: `${digit}`,
    label: `${digit}`,
    shortcut: `${digit}`,
    type: 'Digit'
  }));

/**
 * All operations
 */
export const buttonsData = [
  {
    value: 'divide',
    label: '÷',
    shortcut: '/',
    type: 'Binary',
    work: divide,
    priority: 'HIGH'
  },
  {
    value: 'multiply',
    label: '×',
    shortcut: '*',
    type: 'Binary',
    work: multiply,
    priority: 'HIGH'
  },
  {
    value: 'subtract',
    label: '−',
    shortcut: '-',
    type: 'Binary',
    work: subtract,
    priority: 'LOW'
  },
  {
    value: 'add',
    label: '+',
    shortcut: '+',
    type: 'Binary',
    work: add,
    priority: 'LOW'
  },
  {
    value: 'negate',
    label: '+/-',
    type: 'Unary',
    work: negate
  },
  {
    value: 'percent',
    label: '%',
    shortcut: '%',
    type: 'Unary',
    work: percent
  },
  {
    value: 'equal',
    label: '=',
    shortcut: 'Enter',
    type: 'Equal'
  },
  {
    value: 'dot',
    label: ',',
    type: 'Dot'
  },
  {
    value: 'reset',
    label: 'AC',
    shortcut: 'Escape',
    type: 'Reset'
  },
  ...digits
];

export const availableShortcuts = buttonsData.map(({ shortcut }) => shortcut);

export const getDataByValue = ({ value: targetValue }) => (
  buttonsData.filter(({ value }) => value === targetValue)[0]
);

export const getDataByShortcut = ({ shortcut: targetShortcut }) => (
  buttonsData.filter(({ shortcut }) => shortcut === targetShortcut)[0]
);
