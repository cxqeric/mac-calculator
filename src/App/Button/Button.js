import React from 'react';
import PropTypes from 'prop-types';

import Calculator from '../Calculator/Calculator';
import { getDataByValue } from '../../utils';

class Button extends React.Component {
  static propTypes = {
    value: PropTypes.string.isRequired
  };

  static contextTypes = {
    [Calculator.CALCULATOR_CONTEXT]: PropTypes.shape({
      initial: PropTypes.bool,
      handleInput: PropTypes.func
    })
  };

  handleClick = () => {
    this.context[Calculator.CALCULATOR_CONTEXT].handleInput({ value: this.props.value });
  };

  render() {
    const { value } = this.props;
    const { initial } = this.context[Calculator.CALCULATOR_CONTEXT];
    const buttonData = getDataByValue({ value });

    if (!buttonData) {
      return '';
    }

    const {
      type, label, nonInitialLabel, span
    } = buttonData;

    return (
      <input
        type="button"
        key={value}
        className={`${type} ${type}-${value} span-${span}`}
        onClick={this.handleClick}
        value={!initial && nonInitialLabel ? nonInitialLabel : label}
      />
    );
  }
}


export default Button;
