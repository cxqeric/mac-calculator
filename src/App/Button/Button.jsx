import React from 'react';
import PropTypes from 'prop-types';

import Calculator from '../Calculator';
import { getDataByValue } from '../../utils';

class Button extends React.Component {
  static propTypes = {
    value: PropTypes.string,
    span: PropTypes.string,
    nonInitialLabel: PropTypes.string
  };

  static defaultProps = {
    value: undefined,
    span: undefined,
    nonInitialLabel: undefined
  };

  static contextTypes = {
    [Calculator.CALCULATOR_CONTEXT]: PropTypes.shape({
      initial: PropTypes.bool,
      handleInput: PropTypes.func
    })
  };

  handleClick = () => {
    this.context[Calculator.CALCULATOR_CONTEXT].handleInput({ value: this.props.value });
  };

  render() {
    const { value = '', span = 1, nonInitialLabel } = this.props;
    const { initial } = this.context[Calculator.CALCULATOR_CONTEXT];
    const buttonData = getDataByValue({ value });

    if (!buttonData) {
      return <input type="button" />;
    }

    const { type, label } = buttonData;

    return (
      <input
        type="button"
        key={value}
        className={`${type} ${type}-${value} span-${span}`}
        onClick={this.handleClick}
        value={!initial && nonInitialLabel ? nonInitialLabel : label}
      />
    );
  }
}


export default Button;
