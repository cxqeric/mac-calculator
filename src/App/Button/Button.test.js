import React from 'react';
import { shallow, mount } from 'enzyme';

import Calculator from '../Calculator';
import Button from './Button';
import { getDataByValue } from '../../utils';


describe('Button', () => {
  test('renders empty button', () => {
    const context = { [Calculator.CALCULATOR_CONTEXT]: { initial: true } };
    const wrapper = shallow(<Button />, { context });
    expect(wrapper).toMatchSnapshot();
  });

  describe('button with value', () => {
    test('should render button', () => {
      const context = { [Calculator.CALCULATOR_CONTEXT]: { initial: true } };
      const value = 'add';
      const wrapper = mount(<Button value={value} />, { context });
      expect(wrapper).toMatchSnapshot();
    });

    test('key should be equal to \'value\' prop', () => {
      const context = { [Calculator.CALCULATOR_CONTEXT]: { initial: true } };
      const value = 'add';
      const wrapper = mount(<Button value={value} />, { context });
      expect(wrapper.find('input').key()).toBe(value);
    });

    describe('className', () => {
      const context = { [Calculator.CALCULATOR_CONTEXT]: { initial: true } };
      const value = 'add';
      const wrapper = mount(<Button value={value} />, { context });
      const { type } = getDataByValue({ value });

      test('className should have type', () => {
        expect(wrapper.find('input').hasClass(type)).toBeTruthy();
      });

      test('className should have type-value', () => {
        expect(wrapper.find('input').hasClass(`${type}-${value}`)).toBeTruthy();
      });

      describe('span', () => {
        test('className should have default span to be 1 when there is no \'span\' prop', () => {
          expect(wrapper.find('input').hasClass('span-1')).toBeTruthy();
        });

        test('className should have \'span\' prop', () => {
          expect(wrapper.find('input').hasClass('span-1')).toBeTruthy();
        });
      });
    });

    describe('when prop \'initial\' changes', () => {
      test('value should be the same if a nonInitialLabel is not provided', () => {
        const context = { [Calculator.CALCULATOR_CONTEXT]: { initial: true } };
        const wrapper = shallow(<Button value="reset" />, { context });
        expect(wrapper).toMatchSnapshot();
        wrapper.setContext({ [Calculator.CALCULATOR_CONTEXT]: { initial: false } });
        expect(wrapper).toMatchSnapshot();
      });

      test('value should change if a nonInitialLabel is provided', () => {
        const context = { [Calculator.CALCULATOR_CONTEXT]: { initial: true } };
        const wrapper = shallow(<Button value="reset" nonInitialLabel="C" />, { context });
        expect(wrapper).toMatchSnapshot();
        wrapper.setContext({ [Calculator.CALCULATOR_CONTEXT]: { initial: false } });
        expect(wrapper).toMatchSnapshot();
      });
    });

    test('onclick calls context.handleClick with value parameter', () => {
      const handleInput = jest.fn();
      const context = {
        [Calculator.CALCULATOR_CONTEXT]: {
          initial: true,
          handleInput
        }
      };

      const value = 'add';
      const wrapper = mount(<Button value={value} />, { context });
      wrapper.find('input').simulate('click');
      expect(handleInput).toBeCalledWith({ value });
    });
  });
});
