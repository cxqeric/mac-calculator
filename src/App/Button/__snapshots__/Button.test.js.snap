// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button button with value should render button 1`] = `
<Button
  value="add"
>
  <input
    className="Binary Binary-add span-1"
    key="add"
    onClick={[Function]}
    type="button"
    value="+"
  />
</Button>
`;

exports[`Button button with value when prop 'initial' changes value should be the same if a nonInitialLabel is not provided 1`] = `
<input
  className="Reset Reset-reset span-1"
  key="reset"
  onClick={[Function]}
  type="button"
  value="AC"
/>
`;

exports[`Button button with value when prop 'initial' changes value should be the same if a nonInitialLabel is not provided 2`] = `
<input
  className="Reset Reset-reset span-1"
  key="reset"
  onClick={[Function]}
  type="button"
  value="AC"
/>
`;

exports[`Button button with value when prop 'initial' changes value should change if a nonInitialLabel is provided 1`] = `
<input
  className="Reset Reset-reset span-1"
  key="reset"
  onClick={[Function]}
  type="button"
  value="AC"
/>
`;

exports[`Button button with value when prop 'initial' changes value should change if a nonInitialLabel is provided 2`] = `
<input
  className="Reset Reset-reset span-1"
  key="reset"
  onClick={[Function]}
  type="button"
  value="C"
/>
`;

exports[`Button renders empty button 1`] = `
<input
  type="button"
/>
`;
